src/turbomind/models/llama/LlamaBatch.cc

```diff
@@ -1273,28 +1280,31 @@ auto LlamaBatch::Interrupt(int index, bool force_stop, bool force_end) -> Signal
         TM_LOG_INFO("[Interrupt] slot %d, tokens [%s]", index, ss.str().c_str());
     }
 
-    if (state_->requests[index]->session.end_flag || force_end) {
-        // Sequence is ending this round or a stop request is issued to end it
-        FT_CHECK(sequence_manager_->Erase(state_->requests[index]->id));
-    }
-    else {
-        const int output_len = state_->h_context_length[index];
-        auto&     seq        = *state_->sequences[index];
+    const int output_len = state_->h_context_length[index];
+    auto&     seq        = *state_->sequences[index];
 
-        // Update token IDs
-        seq.tokens.resize(output_len);
+    // Update token IDs
+    seq.tokens.resize(output_len);
 
-        // output_ids is updated & synced in `Finish`
-        const auto output_ids = state_->requests[index]->output_ids.data();
-        std::copy_n(output_ids, output_len, seq.tokens.data());
+    // output_ids is updated & synced in `Finish`
+    const auto output_ids = state_->requests[index]->output_ids.data();
+    std::copy_n(output_ids, output_len, seq.tokens.data());
+    // Cache the generated tokens of the sequence
+    if (!force_stop) {
+        sequence_manager_->CacheGeneration(seq);
+    }
 
-        // Save random state in host memory
-        seq.random_state.resize(sizeof(curandState_t));
-        // This async copy must be synchronized by the caller
-        core::Copy((curandState_t*)state_->curand_state.data() + index, 1, (curandState_t*)seq.random_state.data());
+    // Save random state in host memory
+    seq.random_state.resize(sizeof(curandState_t));
+    // This async copy must be synchronized by the caller
+    core::Copy((curandState_t*)state_->curand_state.data() + index, 1, (curandState_t*)seq.random_state.data());
 
-        // Set unlock flag for corresponding blocks, will be unlocked in the next `Materialize()`
-        sequence_manager_->UpdateAndSetUnlock(seq);
+    // Set unlock flag for corresponding blocks, will be unlocked in the next `Materialize()`
+    sequence_manager_->UpdateAndSetUnlock(seq);
+
+    if (state_->requests[index]->session.end_flag) {
+        // Sequence is ending this round
+        FT_CHECK(sequence_manager_->Erase(state_->requests[index]->id));
```

Comment:
I think we should refer to previous code here. With `state_->requests[index]->session.end_flag=True`, there is no need to output token id and save random_state.